src/Api/Managers/DocumentManager.cs

```diff
@@ -42,6 +49,45 @@ internal EntityResult UpdateDocument(int topicId, int userId, string htmlContent
                 document = new Document(topicId, userId, htmlContent);
                 DbContext.Add(document);
             }
+
+            // document is saved, so now we can parse it
+            var stream = new System.IO.StringReader("<pseudo-root>" + htmlContent + "</pseudo-root>");
+            var xmlReader = XmlReader.Create(stream);
+            var tagInstances = new List<AnnotationTagInstance>();
+
+            try
+            {
+                while (xmlReader.Read())
+                {
+                    if (xmlReader.NodeType != XmlNodeType.Element) continue;
+                    if (!xmlReader.HasAttributes) continue;
+
+                    var tagModelId = int.Parse(xmlReader.GetAttribute("data-tag-model-id"));
+                    var tagInstanceId = int.Parse(xmlReader.GetAttribute("data-tag-id"));
+                    var tagValue = xmlReader.ReadElementContentAsString();
+                    var rx = new Regex("<span[^>]+?data-tag-id=\"" + tagInstanceId + "\".*?>");
+                    var tagPosition = rx.Match(htmlContent).Index;
+
+                    var tagModel = DbContext.AnnotationTags.First(t => t.Id == tagModelId);
+                    var tag = new AnnotationTagInstance(tagModel)
+                    {
+                        IdInDocument = tagInstanceId,
+                        Value = tagValue,
+                        PositionInDocument = tagPosition,
+                        InDocument = document
+                    };
+
+                    tagInstances.Add(tag);
+                }
+                var oldInstances = DbContext.AnnotationTagInstances;
+                DbContext.AnnotationTagInstances.RemoveRange(oldInstances);
```

Comment:
I think the Topic/Document Id is missing. Here you delete all Instanced inside the database.